## Data Cleaning & Transforming

Data Cleaning is an essential step towards making real meaning of any dataset. To make the Company data ready for futher analysis and modelling, data cleaning would be performed.

In [5]:
# import neccessary libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [6]:
# read csv data into a datframe
data = pd.read_csv('../data/Week1_challenge_data_source.csv')
data.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0  9.16456699548519E+015  ...          15854611.0           2501332.0   
1                L77566A  ...          20247395.0          19111729.0   
2                D42335A  ...          19725661.0          14699576.0   
3                T21824A  ...          21388122.0          15146643.0   
4                D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 55 columns]

For better understanding of the columns, exploring the data description [here](https://docs.google.com/spreadsheets/d/1pcNqeUeIph6xAQzlI54KCvi8HM91SUNeeDbdOq3rvbE/edit?ts=60ec0f05#gid=784155610) will help

### Data Understanding/ Exploration

Before proceeding to cleaning the dataset, let us explore what has been stored in it. How many data points present, the column names, datatypes of each colum, descriptive statistics of variables, presence of missing values etc. All these questions needs to be answered.

In [29]:
import sys
sys.path.insert(0,'../scripts/')

In [30]:
from data_information import DataInfo # import data_information script
explorer = DataInfo(data)

In [31]:
# size of data
explorer.get_size()

(150001, 55)

In [32]:
# number of data points
explorer.get_no_datapoints()

 There are 150001 rows and 55 columns presnet in this data


In [33]:
# list the 55 column names
explorer.get_column_names()

Columns in the data are: 


['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Socia

In [35]:
# get information about the data
explorer.get_info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

The data contains 50 numerical datatypes and 5 categorical datatpyes with a memory usage of over 62.9 MB.

In [36]:
# descriptive statistics of data
explorer.get_description()

Descriptive statistics: 


Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082016e+14   
std    2.893173e+18     288.611834     288.097653  8.103762e+04  2.148809e+10   
min    6.917538e+18       0.000000       0.000000  7.142000e+03  2.040471e+14   
25%    7.349883e+18     250.000000     251.000000  5.744050e+04  2.082014e+14   
50%    7.349883e+18     499.000000     500.000000  8.639900e+04  2.082015e+14   
75%    1.304243e+19     749.000000     750.000000  1.324302e+05  2.082018e+14   
max    1.318654e+19     999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847455e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365130e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
count            150000.000000  ...        1.500010e+05        1.500010e+05   
mean              13300.045927  ...        1.163407e+07        1.100941e+07   
std               23971.878541  ...        6.710569e+06        6.345423e+06   
min                   0.000000  ...        5.300000e+01        1.050000e+02   
25%                  43.000000  ...        5.833501e+06        5.517965e+06   
50%                  63.000000  ...        1.161602e+07        1.101345e+07   
75%               19710.750000  ...        1.744852e+07        1.651556e+07   
max              378160.000000  ...        2.325910e+07        2.201196e+07   

       Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
count        1.500010e+05        1.500010e+05       1.500010e+05   
mean         1.162685e+07        1.100175e+07       4.220447e+08   
std          6.725218e+06        6.359490e+06       2.439675e+08   
min          4.200000e+01        3.500000e+01       2.516000e+03   
25%          5.777156e+06        5.475981e+06       2.104733e+08   
50%          1.164222e+07        1.099638e+07       4.234081e+08   
75%          1.747048e+07        1.650727e+07       6.331742e+08   
max          2.325919e+07        2.201196e+07       8.434419e+08   

       Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
count       1.500010e+05      1.500010e+05      1.500010e+05   
mean        8.288398e+06      4.211005e+08      8.264799e+06   
std         4.782700e+06      2.432050e+08      4.769004e+06   
min         5.900000e+01      3.290000e+03      1.480000e+02   
25%         4.128476e+06      2.101869e+08      4.145943e+06   
50%         8.291208e+06      4.218030e+08      8.267071e+06   
75%         1.243162e+07      6.316918e+08      1.238415e+07   
max         1.655879e+07      8.434425e+08      1.655882e+07   

       Total UL (Bytes)  Total DL (Bytes)  
count      1.500000e+05      1.500000e+05  
mean       4.112121e+07      4.546434e+08  
std        1.127639e+07      2.441429e+08  
min        2.866892e+06      7.114041e+06  
25%        3.322201e+07      2.431068e+08  
50%        4.114331e+07      4.558411e+08  
75%        4.903424e+07      6.657055e+08  
max        7.833131e+07      9.029696e+08  

[8 rows x 50 columns]

There seems to be some outlier values present in some columns such as Dur.(ms), IMSI etc through comparism of its mean and standard deviation. This can be justified further in the later causes of this notebook.

In [37]:
# check for total number of missing values
explorer.get_total_missing_values()

The number of missing value(s): 1031392


In [38]:
explorer.get_percentage_missing_values()

There are 12.5 % missing values.


Wow, that is quite a whole lot of missing values. Let us check for columns with these missing values.

In [39]:
# columns with missing values
explorer.get_missing_columns()

Columns having missing value(s): Index(['Bearer Id', 'Start', 'Start ms', 'End', 'End ms', 'Dur. (ms)', 'IMSI',
       'MSISDN/Number', 'IMEI', 'Last Location Name', 'Avg RTT DL (ms)',
       'Avg RTT UL (ms)', 'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)',
       'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)',
       'DL TP < 50 Kbps (%)', '50 Kbps < DL TP < 250 Kbps (%)',
       '250 Kbps < DL TP < 1 Mbps (%)', 'DL TP > 1 Mbps (%)',
       'UL TP < 10 Kbps (%)', '10 Kbps < UL TP < 50 Kbps (%)',
       '50 Kbps < UL TP < 300 Kbps (%)', 'UL TP > 300 Kbps (%)',
       'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Activity Duration DL (ms)',
       'Activity Duration UL (ms)', 'Dur. (ms).1', 'Handset Manufacturer',
       'Handset Type', 'Nb of sec with 125000B < Vol DL',
       'Nb of sec with 1250B < Vol UL < 6250B',
       'Nb of sec with 31250B < Vol DL < 125000B',
       'Nb of sec with 37500B < Vol UL',
       'Nb of sec with 6250B < Vol DL < 31250B',
       'Nb of se

In [42]:
# find percentage of missing values columns
explorer.get_percentage_missing_columns()

Missing Count  %Missing
ColumnName                                                       
Bearer Id                                           991       1.0
Start                                                 1       0.0
Start ms                                              1       0.0
End                                                   1       0.0
End ms                                                1       0.0
Dur. (ms)                                             1       0.0
IMSI                                                570       0.0
MSISDN/Number                                      1066       1.0
IMEI                                                572       0.0
Last Location Name                                 1153       1.0
Avg RTT DL (ms)                                   27829      19.0
Avg RTT UL (ms)                                   27812      19.0
Avg Bearer TP DL (kbps)                               1       0.0
Avg Bearer TP UL (kbps)                               1       0.0
TCP DL Retrans. Vol (Bytes)                       88146      59.0
TCP UL Retrans. Vol (Bytes)                       96649      64.0
DL TP < 50 Kbps (%)                                 754       1.0
50 Kbps < DL TP < 250 Kbps (%)                      754       1.0
250 Kbps < DL TP < 1 Mbps (%)                       754       1.0
DL TP > 1 Mbps (%)                                  754       1.0
UL TP < 10 Kbps (%)                                 792       1.0
10 Kbps < UL TP < 50 Kbps (%)                       792       1.0
50 Kbps < UL TP < 300 Kbps (%)                      792       1.0
UL TP > 300 Kbps (%)                                792       1.0
HTTP DL (Bytes)                                   81474      54.0
HTTP UL (Bytes)                                   81810      55.0
Activity Duration DL (ms)                             1       0.0
Activity Duration UL (ms)                             1       0.0
Dur. (ms).1                                           1       0.0
Handset Manufacturer                                572       0.0
Handset Type                                        572       0.0
Nb of sec with 125000B < Vol DL                   97538      65.0
Nb of sec with 1250B < Vol UL < 6250B             92894      62.0
Nb of sec with 31250B < Vol DL < 125000B          93586      62.0
Nb of sec with 37500B < Vol UL                   130254      87.0
Nb of sec with 6250B < Vol DL < 31250B            88317      59.0
Nb of sec with 6250B < Vol UL < 37500B           111843      75.0
Nb of sec with Vol DL < 6250B                       755       1.0
Nb of sec with Vol UL < 1250B                       793       1.0
Social Media DL (Bytes)                               0       0.0
Social Media UL (Bytes)                               0       0.0
Google DL (Bytes)                                     0       0.0
Google UL (Bytes)                                     0       0.0
Email DL (Bytes)                                      0       0.0
Email UL (Bytes)                                      0       0.0
Youtube DL (Bytes)                                    0       0.0
Youtube UL (Bytes)                                    0       0.0
Netflix DL (Bytes)                                    0       0.0
Netflix UL (Bytes)                                    0       0.0
Gaming DL (Bytes)                                     0       0.0
Gaming UL (Bytes)                                     0       0.0
Other DL (Bytes)                                      0       0.0
Other UL (Bytes)                                      0       0.0
Total UL (Bytes)                                      1       0.0
Total DL (Bytes)                                      1       0.0

Let us drop columns with more than 30% missing values and then fill up the other columns.

In [43]:
# check for duplicate rows
explorer.get_duplicates()

Empty DataFrame
Columns: [Bearer Id, Start, Start ms, End, End ms, Dur. (ms), IMSI, MSISDN/Number, IMEI, Last Location Name, Avg RTT DL (ms), Avg RTT UL (ms), Avg Bearer TP DL (kbps), Avg Bearer TP UL (kbps), TCP DL Retrans. Vol (Bytes), TCP UL Retrans. Vol (Bytes), DL TP < 50 Kbps (%), 50 Kbps < DL TP < 250 Kbps (%), 250 Kbps < DL TP < 1 Mbps (%), DL TP > 1 Mbps (%), UL TP < 10 Kbps (%), 10 Kbps < UL TP < 50 Kbps (%), 50 Kbps < UL TP < 300 Kbps (%), UL TP > 300 Kbps (%), HTTP DL (Bytes), HTTP UL (Bytes), Activity Duration DL (ms), Activity Duration UL (ms), Dur. (ms).1, Handset Manufacturer, Handset Type, Nb of sec with 125000B < Vol DL, Nb of sec with 1250B < Vol UL < 6250B, Nb of sec with 31250B < Vol DL < 125000B, Nb of sec with 37500B < Vol UL, Nb of sec with 6250B < Vol DL < 31250B, Nb of sec with 6250B < Vol UL < 37500B, Nb of sec with Vol DL < 6250B, Nb of sec with Vol UL < 1250B, Social Media DL (Bytes), Social Media UL (Bytes), Google DL (Bytes), Google UL (Bytes), Email DL (Bytes), Email UL (Bytes), Youtube DL (Bytes), Youtube UL (Bytes), Netflix DL (Bytes), Netflix UL (Bytes), Gaming DL (Bytes), Gaming UL (Bytes), Other DL (Bytes), Other UL (Bytes), Total UL (Bytes), Total DL (Bytes)]
Index: []

[0 rows x 55 columns]

The data contains no duplicate rows

### Dealing with missing values

In [44]:
# drop columns with more than 30% missing values
data = data.drop(['TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)', 'HTTP DL (Bytes)', 
                  'HTTP UL (Bytes)', 'Nb of sec with 125000B < Vol DL', 'Nb of sec with 1250B < Vol UL < 6250B',
                  'Nb of sec with 31250B < Vol DL < 125000B','Nb of sec with 37500B < Vol UL',
                  'Nb of sec with 6250B < Vol DL < 31250B','Nb of sec with 6250B < Vol UL < 37500B'], axis=1)

In [45]:
data.shape

(150001, 45)

check if numeric missing values is skew, and fill missing values with mean value else median value if skew

In [106]:
# function to fill numeric values
def fill_numeric_values(df, acceptable_skewness: float = 5.0):
        #df_skew_data = df
        df_skew = df.skew(axis=0, skipna=True)
        for i in df_skew.index:
            if(df_skew[i] < acceptable_skewness and df_skew[i] > (acceptable_skewness * -1)):
                value = df[i].mean()
                df[i].fillna(value, inplace=True)
            else:
                value = df[i].median()
                df[i].fillna(value, inplace=True)

        return df

In [107]:
fill_numeric_values(data)

Bearer Id            Start  Start ms              End     End ms  \
0       1.311448e+19   4/4/2019 12:01  770.0000  4/25/2019 14:35  662.00000   
1       1.311448e+19   4/9/2019 13:04  235.0000   4/25/2019 8:15  606.00000   
2       1.311448e+19   4/9/2019 17:42    1.0000  4/25/2019 11:58  652.00000   
3       1.311448e+19   4/10/2019 0:31  486.0000   4/25/2019 7:36  171.00000   
4       1.311448e+19  4/12/2019 20:10  565.0000  4/25/2019 10:40  954.00000   
...              ...              ...       ...              ...        ...   
149996  7.277826e+18   4/29/2019 7:28  451.0000   4/30/2019 6:02  214.00000   
149997  7.349883e+18   4/29/2019 7:28  483.0000  4/30/2019 10:41  187.00000   
149998  1.311448e+19   4/29/2019 7:28  283.0000  4/30/2019 10:46  810.00000   
149999  1.311448e+19   4/29/2019 7:28  696.0000  4/30/2019 10:40  327.00000   
150000  1.013887e+19              NaN  499.1882              NaN  498.80088   

           Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1.823652e+06  2.082014e+14   3.366496e+10  3.552121e+13   
1       1.365104e+06  2.082019e+14   3.368185e+10  3.579401e+13   
2       1.361762e+06  2.082003e+14   3.376063e+10  3.528151e+13   
3       1.321509e+06  2.082014e+14   3.375034e+10  3.535661e+13   
4       1.089009e+06  2.082014e+14   3.369980e+10  3.540701e+13   
...              ...           ...            ...           ...   
149996  8.123000e+04  2.082022e+14   3.365069e+10  3.548311e+13   
149997  9.797000e+04  2.082019e+14   3.366345e+10  3.566051e+13   
149998  9.824900e+04  2.082017e+14   3.362189e+10  3.572121e+13   
149999  9.791000e+04  2.082021e+14   3.361962e+10  8.618620e+13   
150000  1.046086e+05  2.082015e+14   3.366371e+10  4.847455e+13   

           Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0       9.16456699548519E+015  ...        1.585461e+07        2.501332e+06   
1                     L77566A  ...        2.024740e+07        1.911173e+07   
2                     D42335A  ...        1.972566e+07        1.469958e+07   
3                     T21824A  ...        2.138812e+07        1.514664e+07   
4                     D88865A  ...        1.525938e+07        1.896287e+07   
...                       ...  ...                 ...                 ...   
149996                D20434A  ...        1.619167e+07        1.176343e+07   
149997                D10223C  ...        1.387723e+07        8.288284e+06   
149998                T51102A  ...        2.266051e+07        1.855903e+06   
149999                L88342B  ...        8.817106e+06        8.305402e+06   
150000                    NaN  ...        1.163407e+07        1.100941e+07   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0             8.198936e+06        9.656251e+06       2.780823e+08   
1             1.833841e+07        1.722713e+07       6.087501e+08   
2             1.758779e+07        6.163408e+06       2.295846e+08   
3             1.399465e+07        1.097942e+06       7.995382e+08   
4             1.712458e+07        4.152180e+05       5.277072e+08   
...                    ...                 ...                ...   
149996        1.788370e+07        1.967816e+07       5.266097e+08   
149997        1.935015e+07        2.129315e+07       6.268931e+08   
149998        9.963942e+06        5.065760e+06       5.535395e+08   
149999        3.322253e+06        1.317259e+07       3.525370e+08   
150000        1.162685e+07        1.100175e+07       4.220447e+08   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0            1.434415e+07      1.717444e+08      8.814393e+06   
1            1.170709e+06      5.269042e+08      1.505514e+07   
2            3.956300e+05      4.106926e+08      4.215763e+06   
3            1.084972e+07      7.490399e+08      1.279728e+07   
4            3.529801e+06      5.507095e+08      1.391032e+07   
...                   ...               ...               ...   
149996       9.197207e+06      3.264

In [108]:
# confirm columns with missing values
print(data.columns[data.isnull().any()])

Index(['Start', 'End', 'Last Location Name', 'Handset Manufacturer',
       'Handset Type'],
      dtype='object')


In [110]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 45 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Bearer Id                       150001 non-null  float64
 1   Start                           150000 non-null  object 
 2   Start ms                        150001 non-null  float64
 3   End                             150000 non-null  object 
 4   End ms                          150001 non-null  float64
 5   Dur. (ms)                       150001 non-null  float64
 6   IMSI                            150001 non-null  float64
 7   MSISDN/Number                   150001 non-null  float64
 8   IMEI                            150001 non-null  float64
 9   Last Location Name              148848 non-null  object 
 10  Avg RTT DL (ms)                 150001 non-null  float64
 11  Avg RTT UL (ms)                 150001 non-null  float64
 12  Avg Bearer TP DL

Fill Start and End columns missing values with forward and backward fill.

In [125]:
#forward - start and end 

def fill_values(df, col, ffill: bool = True, bfill: bool = False):
    if(ffill == True and bfill == True):
        df[col] = df[col].fillna(method='ffill')
        df[col] = df[col].fillna(method='bfill')
    
    elif(ffill == True and bfill == False):
        df[col] = df[col].fillna(method='ffill')
    
    elif(ffill == False and bfill == True):
        df[col] = df[col].fillna(method='bfill')
        
    else:
        df[col] = df[col].fillna(method='bfill')
        df[col] = df[col].fillna(method='ffill')
        
    return df[col]

In [126]:
df['Start'] = fill_values(df = data, col = 'Start')
df['End'] = fill_values(df = data, col = 'End')

In [127]:
data.columns[data.isnull().any()]

Index(['Last Location Name', 'Handset Manufacturer', 'Handset Type'], dtype='object')

Fill remaining Categorical missing values with mode

In [128]:
# fill 'race' column with mode 
data['Last Location Name'] = data['Last Location Name'].fillna(data['Last Location Name'].mode()[0])
data['Handset Manufacturer'] = data['Handset Manufacturer'].fillna(data['Handset Manufacturer'].mode()[0])
data['Handset Type'] = data['Handset Type'].fillna(data['Handset Type'].mode()[0])

In [132]:
# confirm if there are any missing values
print('There are {}'.format(data.isnull().sum().sum()))

There are 0 missing value(s)


All missing values dealt with!!

### Sub-tasks Analysis

* Top 10 handsets used by the customers:

In [133]:
# calculate the top 10 most used handsets
handsets = data['Handset Type'].value_counts()[:10]
handsets

Huawei B528S-23A                20324
Apple iPhone 6S (A1688)          9419
Apple iPhone 6 (A1586)           9023
undefined                        8987
Apple iPhone 7 (A1778)           6326
Apple iPhone Se (A1723)          5187
Apple iPhone 8 (A1905)           4993
Apple iPhone Xr (A2105)          4568
Samsung Galaxy S8 (Sm-G950F)     4520
Apple iPhone X (A1901)           3813
Name: Handset Type, dtype: int64

* Top 3 handset manufacturers

In [134]:
# identify the top 3 handset manufacturers
handset_manu = data['Handset Manufacturer'].value_counts()[:3]
handset_manu

Apple      60137
Samsung    40839
Huawei     34423
Name: Handset Manufacturer, dtype: int64

* Top 5 Handset per top 3 Handset Manufacturer

In [74]:
# identify the top 5 Handset per top 3 Handset Manufacturer
df = data.loc[:,['Handset Type','Handset Manufacturer']] # create a dataframe containg all rows from the actual data and select ahndset type and manufacturer columns
value = df.groupby(['Handset Manufacturer', 'Handset Type']).size() # groupby the columns
df_list = pd.Series(dtype='object') # create a pandas series and specify the object datatype

for h in handset_manu.index:
    df_list = df_list.append(value[h]) # select the top 3 manufacurers
df_list.sort_values(ascending=False)[:5] # sort by top 5

Huawei B528S-23A           19752
Apple iPhone 6S (A1688)     9419
Apple iPhone 6 (A1586)      9023
Apple iPhone 7 (A1778)      6326
Apple iPhone Se (A1723)     5187
dtype: int64

#### Interpretation and Recommendation

Analysis from the top 10 most used handsets by customers, shows Huawei, Apple and Samsung models widely used. The top 3 maufacturers of this handsets as expected are the Apple, Samsung and Huawei companies respectfully. The most used handset is Hwawei B528S-23A with 19752 users followed by Apple iPhhones 6S (A1688), 6 (A1586), 7 (A1778) and Se (A1723) ascendingly. 

Since the Huawei B528S-23A handset is most used by customers, making promotions, expanding its features, selling the same phone and its newer version are profitable. It can  also be infered that iPhones are the next widely used handsets and similar marketing ideas can implemented as well.

### Overview of the users’ behavior

Aggregate per user ('MSISDN/Number'), the; number of xDR sessions, session duration, the total download (DL) and upload (UL) data, the total data volume (in Bytes) during this session for each application. 

In [75]:
#pd.set_option('display.float_format', '{:.2f}'.format)

To_Do:
* check for outliers
* EDA
* update README

## Data Exploration